### DISCUSSION TITLE
Server connection lost

### DISCUSSION THREAD

**Discussion by dveretennikov1** [outline/outline#3316] [Self Hosting]
I've got very strange error, when I'm trying to edit any doc 
![Screenshot 2022-03-31 195557](https://user-images.githubusercontent.com/102303381/161109400-96b81d5b-7ed8-45b4-b108-822a0f03f5dc.png)
I have no idea what that meaning

Logs
{"level":"warn","message":"Enforced https was disabled with FORCE_HTTPS env variable"}
{"label":"lifecycle","level":"info","message":"Listening on http://localhost:3000"}
{"name":"views.create","modelId":null,"label":"processor","level":"info","message":"backlinks processing views.create"}
{"name":"views.create","modelId":null,"label":"processor","level":"info","message":"debouncer processing views.create"}
{"name":"views.create","modelId":null,"label":"processor","level":"info","message":"imports processing views.create"}
{"name":"views.create","modelId":null,"label":"processor","level":"info","message":"exports processing views.create"}
{"name":"views.create","modelId":null,"label":"processor","level":"info","message":"notifications processing views.create"}
{"name":"views.create","modelId":null,"label":"processor","level":"info","message":"revisions processing views.create"}
{"name":"views.create","modelId":null,"label":"processor","level":"info","message":"slack processing views.create"}
{"name":"views.create","modelId":null,"label":"processor","level":"info","message":"backlinks processing views.create"}
{"name":"views.create","modelId":null,"label":"processor","level":"info","message":"debouncer processing views.create"}
{"name":"views.create","modelId":null,"label":"processor","level":"info","message":"imports processing views.create"}
{"name":"views.create","modelId":null,"label":"processor","level":"info","message":"exports processing views.create"}
{"name":"views.create","modelId":null,"label":"processor","level":"info","message":"notifications processing views.create"}
{"name":"views.create","modelId":null,"label":"processor","level":"info","message":"revisions processing views.create"}
{"name":"views.create","modelId":null,"label":"processor","level":"info","message":"slack processing views.create"}
{"name":"views.create","modelId":null,"label":"processor","level":"info","message":"backlinks processing views.create"}
{"name":"views.create","modelId":null,"label":"processor","level":"info","message":"debouncer processing views.create"}
{"name":"views.create","modelId":null,"label":"processor","level":"info","message":"imports processing views.create"}
{"name":"views.create","modelId":null,"label":"processor","level":"info","message":"exports processing views.create"}
{"name":"views.create","modelId":null,"label":"processor","level":"info","message":"notifications processing views.create"}
{"name":"views.create","modelId":null,"label":"processor","level":"info","message":"revisions processing views.create"}
{"name":"views.create","modelId":null,"label":"processor","level":"info","message":"slack processing views.create"}

**Comment by almereyda** [comment]
Is it possible that you have collaborative editing enabled in your instance?

Then the ideas expressed in the comment at https://github.com/outline/outline/discussions/3300#discussioncomment-2460691 might help you to debug the situation any further:

> Which operating system are you running, how do you deploy Outline and the associated services (Redis, Postgres, S3), and especially which kind of setup are you using to provide TLS termination and routing of web requests?
>
> I suspect this is due to a Websocket connection, which is not properly proxied.
>
> Could you also check the network tab of your browser's developer tools, if there is something that errors out? Ideally you should see HTTP 101 GET requests to the `wss://` endpoints that open a channel for synchronisation messages, in my case visible in the "Response" tab of a selected transmission.

**Reply by dveretennikov1** [reply]
Exactly, I turned on collaborative and forget about it.
I use docker containers for all services and nginx as proxy

what else should I proxy with nginx?
now I use only this, for outline and minio

proxy_pass http://172.17.0.1:3000
proxy_pass http://172.17.0.1:9000

developers console has an errors
![Screenshot 2022-04-01 123257](https://user-images.githubusercontent.com/102303381/161237142-498144c5-c925-4a05-8e95-fc08eeaac36a.png)

**Reply by almereyda** [reply]
Are you sure both Outline and Minio proxy rules should point to the same container, but different ports? It feels there might be a cause for a regression, or some other deficiencies with Nginx as a Websocket proxy.

Ref.:

- https://www.nginx.com/blog/websocket-nginx/

**Reply by dveretennikov1** [reply]
Sorry, my mistake, I shortened config. full nginx config
outline.conf
``` server {
  server_name outline.xxx.xxx www.outline.xxx.xxx;
  client_max_body_size 100m;


  location / {

  proxy_set_header X-Real-IP $remote_addr;
   proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
   proxy_set_header X-Forwarded-Proto $scheme;
   proxy_set_header Host $http_host;

   proxy_connect_timeout 300;

   proxy_http_version 1.1;
   proxy_set_header Connection "";
   chunked_transfer_encoding off;


    proxy_pass http://172.17.0.1:3000;
  }


    listen 443 ssl;
    ssl_certificate /etc/letsencrypt/live/outline.xxx.xxx/fullchain.pem; 
    ssl_certificate_key /etc/letsencrypt/live/outline.xxx.xxx/privkey.pem; 
    include /etc/letsencrypt/options-ssl-nginx.conf; 
    ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem; 

}





server {
    if ($host = www.outline.xxx.xxx) {
        return 301 https://$host$request_uri;
    } 

    if ($host = outline.xxx.xxx) {
        return 301 https://$host$request_uri;
    } 


  listen        80;
  server_name outline.xxx.xxx www.outline.xxx.xxx;
    return 404; 

}

```
minio.conf

```
server {

 server_name outlinedata.xxx.xxx www.outlinedata.xxx.xxx;

 ignore_invalid_headers off;
 client_max_body_size 0;
 proxy_buffering off;

 location / {
   proxy_set_header X-Real-IP $remote_addr;
   proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
   proxy_set_header X-Forwarded-Proto $scheme;
   proxy_set_header Host $http_host;

   proxy_connect_timeout 300;

   proxy_http_version 1.1;
   proxy_set_header Connection "";
   chunked_transfer_encoding off;

   proxy_pass http://172.17.0.1:9000; # If you are using docker-compose this would be the hostname i.e. minio
 }

    listen 443 ssl; 
    ssl_certificate /etc/letsencrypt/live/outlinedata.xxx.xxx/fullchain.pem; 
    ssl_certificate_key /etc/letsencrypt/live/outlinedata.xxx.xxx/privkey.pem; 
    include /etc/letsencrypt/options-ssl-nginx.conf; 
    ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem; 


}


server {
    if ($host = www.outlinedata.xxx.xxx) {
        return 301 https://$host$request_uri;
    } # managed by Certbot


    if ($host = outlinedata.xxx.xxx) {
        return 301 https://$host$request_uri;
    } # managed by Certbot


 listen 80;
 server_name outlinedata.xxx.xxx www.outlinedata.xxx.xxx;
    return 404; # managed by Certbot

}

**Reply by almereyda** [reply]
Hi @dveretennikov1 thank you for posting and redacting the details. Since these snippets are very hard to read, could you please try to encapsulete corresponding sections within triple backticks, as described in https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code? Else I fear I won't be able to decypher the above.

**Reply by dveretennikov1** [reply]
done

**Reply by almereyda** [reply]
Hi, thank you for the formatting changes! That made it much easier to read your document.

Have you had a look at the page that I linked at https://www.nginx.com/blog/websocket-nginx/ ?

It describes how to set the Nginx configuration, in so it supports upgrading an HTTP connection to a Websocket connection.

Read through that document and apply the settings as explained there. When you have verified your changed configuration, please report back here with the results. We can then further debug your settings, or congratulate you on the working setup.

But it is necessary that you understand what is happening here around the subject of Websockets, and know how to configure them properly, before any support can be given from the application side.

Also note most of my contributions here are a voluntary effort, so it would feel kind if you followed my suggestions and investigated them on your own, before pasting your configuration. In your example above, there is no mention of

```nginx
    map $http_upgrade $connection_upgrade {
        default Upgrade;
        ''      close;
    }
```

nor

```nginx
            proxy_http_version 1.1;
            proxy_set_header Upgrade $http_upgrade;
            proxy_set_header Connection $connection_upgrade;
```

which made me wonder if you had noticed the link to https://www.nginx.com/blog/websocket-nginx/ before.

**Comment by dveretennikov1** [comment]
I was absolutely shure I copied configs from here
https://app.getoutline.com/share/770a97da-13e5-401e-9f8a-37949c19f97e/doc/nginx-6htaRboR57
and there were no any lines like
proxy_set_header Connection $connection_upgrade;
Maybe I copied config from different source, my mistake. Now I added your recommended lines and everything works
Thanks a lot

**Reply by almereyda** [reply]
Glad to hear this worked out!

It seems the document you linked now contains those two lines:

```nginx
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "Upgrade";
```

and has a little typo around

```nginx
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;proxy_set_header Host $host;
```

where an additional line break after the first `;` semicolon could add additional clarity.

/cc @tommoor for documentation chore
